### ISSUE TITLE
Release from specific commit hash

### ISSUE THREAD

**Issue by CormacdeBarra** [semantic-release/semantic-release#1213]
## New feature motivation

In our CI (Jenkins) we are creating the release after a prod deploy. Its the last step in the deploy process. However this means that it is possible that a release could pick up commits that are not included in the release. e.g.

Commit A is merged into master and triggers a stage build (build #1). It may be on stage for a few days and in the meantime someone else merges a Commit B that also triggers a stage build (build #2). There may be cases where Build #1 is promoted to production before build #2 but AFTER Commit B is merged into master. Semantic release would then pick up commit B which is not actually in build #1 and add it to the release notes for the incorrect release. 

Not sure if this is a bug or a feature request

## New feature description
Add the ability to add a commit hash (the latest commit to include) to semantic release so that the boundaries are set on a release. The only way I can see around this is to use a release branch which is not a path I want to take. 

Seems like useful feature to have the ability to supply the commit hash for the release.

## Possible implemention
add a parameter `latestCommit` or similar to https://github.com/semantic-release/semantic-release/blob/master/lib/get-commits.js and modify to use this commit if it exists instead of `HEAD`

I can attempt to PR this if you think its worth it.

**Comment by pvdlg** [comment]
Could you explain your workflow in details?

Especially that part:
> There may be cases where Build #1 is promoted to production before build #2 but AFTER Commit B is merged into master. Semantic release would then pick up commit B which is not actually in build #1 and add it to the release notes for the incorrect release.

It seems you make a difference between build and release. What's the difference? What's a build? What does promoting a build means?

When semantic-release runs it looks at the commits on the current branch, analyze them, decide a release should be published, and if yes generate the release log for those commits. This whole process takes a few minutes top and even if you'd push new commit during that process it would be a problem as semantic-release fetch and reads the commits only once at the.

The commits included in the changelog are always the commits between last release and the HEAD of the current branch. It's impossible that semantic-release would include a commit in the changelog that is not part of the release...where would semantic-release get that commit from anyway?

Finally semantic-release makes releases for every new commits on `master` in the order they get pushed. If you merge commit A to `master`, semantic-release run and publish a release, then when you merge commit B, same thing. The second release will have a higher version number than the first one. If for some reasons you want commit B to be released first then it has to be merged first.
This behavior is extremely important, because it guarantees that your git history match your release history. You don't have to have a change present in `1.1.0` but not in `1.2.0`.

**Comment by CormacdeBarra** [comment]
Thanks for the speedy response.

Let me try explain my situation a little better.

Our workflow looks like this:
1. Develop on a branch and PR into master.
2. Merges into master are picked up by our CI system (Jenkins) and triggers a build. (A build is a pull from master, then run through lint, tests etc and generate release artifacts)
3. Release that build (which is essentially a snapshot of master at that point in time) to UAT.
4. Once a feature is approved on UAT it is promoted to prod. Promoting to prod does not trigger a new build from master. It simply promotes the already tested build from UAT. (This is important as this means potentially not all commits to master are in the release)
5. Once prod deploy is complete. semantic release is run to roll the changelog etc.

What this means is anything that is committed to master between when that build made it to UAT and when its promoted to prod will be picked up by semantic release and included in the changelog when it may not actually be in the build.

> Finally semantic-release makes releases for every new commits on master in the order they get pushed. If you merge commit A to master, semantic-release run and publish a release, then when you merge commit B, same thing. The second release will have a higher version number than the first one. If for some reasons you want commit B to be released first then it has to be merged first.

This depends on when semantic release is run. If semantic release is run on every commit to master then yes. But if its run on every deploy to production when I think is when it should be run then this is possibly not the case. A deply to production could contain multiple commits to master all rolled into one release.

I realise this is an edge case, which will not happen very often if at all, but is better if it can be prevented. 

Essentially our (potential) problem is that a release is not necessarily from `HEAD` which semantic release assumes is always the case.

Does this explain my predicament better?

**Comment by pvdlg** [comment]
semantic-release is opinionated in the sense that we consider that anything that is pushed on the release branch (`master`) in your case is ready to be released and as such should be published right away.

If your code is not ready to be released then it shouldn't be merged to `master`. So you should do your UAT on another branch, and merge that branch to `master` when it's ready to release.

**Comment by chris-tremblay** [comment]
This is the exact scenario I am running into at our company as well.
